x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview_header/engine_overview_header.test.tsx

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import { mount } from 'enzyme';
```

Comment:
I use mount over shallow in most tests for two reasons...

1. Since we use EUI components it would require a lot of diving, which starts to make these tests a bit more brittle, because any UI tweak would also require cause us to adjust our dive chain. Using mount I just mount the whole thing and pick out interesting elements to test using `data-test-subj`.

2. In order to execute `useContext`, we need to use `mount`, afaict.

I this also generally lets us mock less and test bigger unit chunks.

Reply:
My wariness of using `mount()` over `shallow()` stems primarily from my past experiences writing hundreds of Jest tests for a large UI library. The primary dev impact I recall is one of speed - Jest tests are simply much faster (for obvious reasons) with `shallow()`. This only comes into play when we have a larger library with a lot of tests, which we don't have now - but we eventually will once AS moves fully over.

That being said, I accept that is a pre-optimization we don't have to worry about yet, that can easily wait until post-MVP, and you're right that useContext is far more annoying to do in `shallow()` vs. `mount()` (it's possible from my quick research, but it requires some extra setup that we're not doing).

I am slightly more concerned about point 1 - again, this is from past experience with writing tests with 3rd party libs (or even 1st party deps that we'd written ourselves) - but I tend to find that `mount()` tempts people into the trap of diving into a component that doesn't belong to us and that we shouldn't depend upon for tests. From a high level point of view for a unit test, all we care about is that a component named "EuiButton" rendered, and that the correct props were passed to it - not anything that's inside it (that belongs to the unit test for EuiButton itself, which EUI is in charge of).

For example, if we wrote something like `wrapper.find('[data-test-subj="someStringSetByEui"]'` - that's now fragile and causes failures if a 3rd party lib changes without warning. We didn't do so in this test, which is great, but I find that `shallow()` generally tends to discourage / makes it much harder to write code that relies on 3rd party deps by forcing devs to manually dive into components rather than just having everything readily available.

Sorry, all that was a lot of words :) TL;DR: I'm definitely not against moving forward with `mount()`, but especially with performance considerations down the road, I'd like us to be open to moving to `shallow()` as a tech debt item at some point. Does that sound good?

Reply:
I 100% agree with you. I usually default to shallow, and if shallow won't work, I use mount.

One way I've handled the context thing in components is export two versions of a component, the outer component, which wraps an inner component to access context and pass it through as a simple property to the inner component. Then I unit test the InnerComponent.

```
export default OuterComponent = () => {
  const {prop} = useContext(...)
  return <InnerComponent prop={prop}
}

export const InnerComponent = ({prop}) => {
  <div>Hi, I am a {prop}</div>
}
```

Reply:
That was pre-hooks, redux higher-order component days, things are probably different now though.